### PR TITLE
[core][experimental] Add experimental compiled DAG

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This adds an experimental interface that allows you to compile a normal ray.dag to use a faster execution path. The execution path uses ray.experimental.channel to pass data directly between actors instead of going through the normal Ray control plane. See [the REP](https://github.com/ray-project/enhancements/pull/48) for more information.

Several limitations will be addressed in a follow-up PR:
1. `dag.execute` returns a Channel instead of an ObjectRef. Reading from a Channel requires the caller to release the value as soon as they are done with it, to avoid blocking the next DAG execution.
2. Single-node only.
3. Actors can only execute one DAG. Once they begin executing, they cannot execute other tasks and must be killed to stop executing. An actor can have at most one task in a DAG.

Initial results show about 35x speedup compared to normal Ray core on simple single-node DAGs:
```
single-actor DAG calls per second 593.42 +- 32.8
compiled single-actor DAG calls per second 20931.29 +- 675.78
scatter-gather DAG calls per second 248.87 +- 4.06
compiled scatter-gather DAG calls, n=4 actors per second 8896.79 +- 92.42
chain DAG calls, n=4 actors per second 156.48 +- 14.11
compiled chain DAG calls, n=4 actors per second 5738.39 +- 219.31
```

### CODE REVIEW

File: python/ray/dag/class_node.py [hunk 1]
```diff
@@ -85,23 +83,25 @@ def __getattr__(self, method_name: str):
...
     def __str__(self) -> str:
         return get_dag_node_str(self, str(self._body))
 
 
 class _UnboundClassMethodNode(object):
-    def __init__(self, actor: ClassNode, method_name: str):
+    def __init__(self, actor: ClassNode, method_name: str, options: dict):
+        # TODO(sang): Theoretically, We should use weakref cuz it is
+        # a circular dependency but when I used weakref, it fails
+        # because we cannot serialize the weakref.
```
Comment (ericl) [inline python/ray/dag/class_node.py:97]: Explain the concern about circular dependencies here?

File: python/ray/dag/compiled_dag_node.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        self.downstream_node_idxs = set()
+        self.output_channel = None
+
+    @property
+    def args(self):
+        return self.dag_node.get_args()
+
+    @property
+    def num_readers(self):
+        return len(self.downstream_node_idxs)
```
Comment (ericl) [inline python/ray/dag/compiled_dag_node.py:100]: Add types for methods and attributes, here and elsewhere?

File: python/ray/dag/compiled_dag_node.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        if not isinstance(output_node, OutputNode):
+            self.has_single_output = True
+            output_node = OutputNode([output_node])
+            self._add_node(output_node)
+            self.output_task_idx = self.dag_node_to_idx[output_node]
+            # Preprocess one more time so that we have the right output node
+            # now.
+            self._preprocess()
+
+    def _compile(self) -> Tuple[ChannelType, Union[ChannelType, List[ChannelType]]]:
```
Comment (ericl) [inline python/ray/dag/compiled_dag_node.py:230]:
```suggestion
    def _get_or_compile(self) -> Tuple[ChannelType, Union[ChannelType, List[ChannelType]]]:
```

File: python/ray/dag/compiled_dag_node.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+            args: Args to the InputNode.
+            kwargs: Kwargs to the InputNode. Not supported yet.
+
+        Returns:
+            A list of Channels that can be used to read the DAG result.
+        """
+        # These errors should already be caught during compilation, but just in
+        # case.
+        if len(args) != 1:
+            raise ValueError("Compiled DAGs support exactly one InputNode arg")
```
Comment (ericl) [inline python/ray/dag/compiled_dag_node.py:366]:
```suggestion
            raise NotImplementedError("Compiled DAGs support exactly one InputNode arg")
```

File: python/ray/dag/compiled_dag_node.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        if len(kwargs) != 0:
+            raise ValueError("Compiled DAGs do not support kwargs")
```
Comment (ericl) [inline python/ray/dag/compiled_dag_node.py:368]:
```suggestion
            raise NotImplementedError("Compiled DAGs do not support kwargs")
```

File: python/ray/dag/dag_node.py [hunk 6]
```diff
@@ -103,9 +107,21 @@ async def get_object_refs_from_last_execute(self) -> Dict[str, Any]:
     def clear_cache(self):
         self.cache_from_last_execute = {}
 
+    def experimental_compile(self) -> "ray.dag.CompiledDAG":
+        """Compile an accelerated execution path for this DAG. The compiled DAG
+        is cached.
+        """
+        if self._compiled_dag is None:
```
Comment (ericl) [inline python/ray/dag/dag_node.py:114]: Let's not cache this.

File: python/ray/dag/dag_node.py [hunk 7]
```diff
@@ -103,9 +107,21 @@ async def get_object_refs_from_last_execute(self) -> Dict[str, Any]:
     def clear_cache(self):
         self.cache_from_last_execute = {}
 
+    def experimental_compile(self) -> "ray.dag.CompiledDAG":
```
Comment (ericl) [inline python/ray/dag/dag_node.py:110]: Shall we expose the buffer_size_bytes arg here for now?

**Review by ericl — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]

File: python/ray/dag/class_node.py (hunk 1 above)
Comment (edoakes) [reply python/ray/dag/class_node.py:97]: you can add a simple wrapper for `weakref` to address the serialization issue if it's needed. I did the same in a patch to the `pydantic` repo:
https://github.com/pydantic/pydantic/pull/7780

File: python/ray/dag/tests/test_accelerated_dag.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+import ray.cluster_utils
+from ray.dag import InputNode, OutputNode
+from ray.tests.conftest import *  # noqa
+from ray._private.test_utils import wait_for_condition
+
+
+logger = logging.getLogger(__name__)
+
+
+@ray.remote(concurrency_groups={"_ray_system": 1})
```
Comment (edoakes) [inline python/ray/dag/tests/test_accelerated_dag.py:18]: What is the `_ray_system` concurrency group and why's it required here?

Comment (ericl) [inline python/ray/dag/tests/test_accelerated_dag.py:18]:
```suggestion
@ray.remote
```

**Pull request merged**